app/Http/Controllers/Api/ComponentController.php

```
@@ -69,65 +56,77 @@ public function getComponent($id)
-
-            // For every tag, do we need to create it?
-            $componentTags = array_map(function ($taggable) use ($component) {
-                return Tag::firstOrCreate([
-                    'name' => $taggable,
-                ])->id;
-            }, $tags);
-
-            $component->tags()->sync($componentTags);
+        $componentData = Binput::except('tags');
+        $componentData['user_id'] = $auth->user()->id;
+
+        try {
+            $component = Component::create($componentData);
+        } catch (Exception $e) {
+            throw new BadRequestHttpException();
         }
 
-        return $this->item($component);
+        if ($component->isValid()) {
+            if (Binput::has('tags')) {
+                // The component was added successfully, so now let's deal with the tags.
+                $tags = preg_split('/ ?, ?/', Binput::get('tags'));
+
+                // For every tag, do we need to create it?
+                $componentTags = array_map(function ($taggable) use ($component) {
+                    return Tag::firstOrCreate([
+                        'name' => $taggable,
+                    ])->id;
+                }, $tags);
+
+                $component->tags()->sync($componentTags);
+            }
+
+            return $this->item($component);
+        }
+
+        throw new BadRequestHttpException();
     }
 
     /**
      * Update an existing component.
      *
-     * @param int $id
+     * @param \CachetHQ\Cachet\Models\Componet $component
      *
      * @return \CachetHQ\Cachet\Models\Component
      */
-    public function putComponent($id)
+    public function putComponent(Component $component)
     {
-        $component = $this->component->update($id, Binput::except('tags'));
+        $component->update(Binput::except('tags'));
 
-        if (Binput::has('tags')) {
-            $tags = preg_split('/ ?, ?/', Binput::get('tags'));
+        if ($component->isValid('updating')) {
```

Better to throw the exception first, then avoid the nesting.